lib/spack/spack/config.py

```diff
@@ -883,13 +884,14 @@ def add(fullpath, scope=None):
             existing = get_valid_type(path)
 
             # construct value from this point down
-            value = syaml.load_config(components[-1])
             for component in reversed(components[idx + 1 : -1]):
                 value = {component: value}
             break
 
+    if override:
+        path += "::"
+
     if has_existing_value:
-        path, _, value = fullpath.rpartition(":")
         value = syaml.load_config(value)
```

Comment:
This second `value =` should not be required: the one you added at line 866 should be enough. I'm not clear if that is what's causing the test failures though.